app/models/concerns/orchestration/dhcp.rb

```diff
@@ -19,58 +19,61 @@ def dhcp?
         !subnet.nil? && subnet.dhcp? && SETTINGS[:unattended] && (!provision? || operatingsystem.present?)
   end
 
-  def dhcp_record
-    return unless dhcp? || @dhcp_record
-    handle_validation_errors do
-      @dhcp_record ||= (provision? && jumpstart?) ? Net::DHCP::SparcRecord.new(dhcp_attrs) : Net::DHCP::Record.new(dhcp_attrs)
+  def dhcp_records
+    return {} unless dhcp?
+    @dhcp_records ||= mac_addresses_for_provisioning.map do |record_mac|
+      build_dhcp_record(record_mac)
     end
   end
 
   def reset_dhcp_record_cache
-    @dhcp_record = nil
+    @dhcp_records = nil
   end
 
   def rebuild_dhcp
-    if dhcp?
-      del_dhcp_safe
-      begin
-        set_dhcp
-      rescue => e
-        Foreman::Logging.exception "Failed to rebuild DHCP record for #{name}, #{ip}", e, :level => :error
-        false
-      end
-    else
+    unless dhcp?
       logger.info "DHCP not supported for #{name}, #{ip}, skipping orchestration rebuild"
-      true
+      return true
+    end
+
+    del_dhcp_safe
+    begin
+      set_dhcp
+    rescue => e
+      Foreman::Logging.exception "Failed to rebuild DHCP record for #{name}, #{ip}", e, :level => :error
+      false
     end
   end
 
   protected
 
   def del_dhcp_safe
-    if dhcp_record
-      begin
-        del_dhcp
-      rescue => e
-        Foreman::Logging.exception "Proxy failed to delete DHCP record for #{name}, #{ip}", e, :level => :error
-      end
+    return unless dhcp_records.any?
```

Comment:
I believe you can move this call `dhcp_records.any?` to `del_dhcp` so that it looks like:

```
   def del_dhcp
    return true unless dhcp_records.any?
    dhcp_records.all? { |record| record.destroy }
  end
```

Then you can remove it from here